source/development_process/index.rst

```diff
@@ -0,0 +1,958 @@
+  was forgotten, first add any files which were changed or forgotten, then run
+  ``git commit --amend``.
+* To add part of a file, use ``git add -p`` and follow the interactive
+  instructions.
+* To delete a remote branch named ``branch``, run ``git push origin :branch``
+* To get a pretty view of git history, run ``git log --graph --all``.
+  Some developers alias this to ``git-net``.
+
+
+Writing the Chef Cookbook
+-------------------------
+
+Writing a chef cookbook is very similar to writing the rest of the project. The
+Lab has internal resources explaining how to write a cookbook, so this is just
+a summary. A cookbook describes the desired state of a server. A cookbook may
+have several recipes for different operating systems or for different parts of
+the servers configuration. Cookbooks can use pieces of code from other
+cookbooks called resources. Here is an abridged version of the What's Fresh
+coobook's default recipe:
+
+.. code:: ruby
+
+	directory node['whats_fresh']['config_dir'] do
+	  owner node['whats_fresh']['venv_owner']
+	  group node['whats_fresh']['venv_group']
+	  recursive true
+	end
+
+	python_webapp 'whats_fresh' do
+	  create_user true
+	  path node['whats_fresh']['application_dir']
+	  owner node['whats_fresh']['venv_owner']
+	  group node['whats_fresh']['venv_group']
+
+	  repository node['whats_fresh']['repository']
+	  revision node['whats_fresh']['git_branch']
+
+	  config_template 'config.yml.erb'
+	  config_destination "#{node['whats_fresh']['config_dir']}/config.yml"
+	  django_migrate true
+	  django_collectstatic true
+	  interpreter 'python2.7'
+
+	  gunicorn_port node['whats_fresh']['gunicorn_port']
+	end
+
+	nginx_app 'whats_fresh' do
+	  template 'whats_fresh.conf.erb'
+	  cookbook 'whats-fresh'
+	end
+
+This recipe uses several resources, including ``python_webapp``, ``nginx_app``,
+and ``directory``. The section ``node['whats_fresh']['application_dir']`` is
+just like a python dictionary with a specific configuration value, often a
+string.
+
+
+The Architecture of Chef, Abridged
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Chef has several parts, a central server which holds all of the cookbooks, and
+a collection of nodes, which are managed by the chef server. Each node has
+attributes which are essentially a hash or a dictionary of configuration values
+and keys. Each node has a list of roles it serves, like web server or database
+server, and a run list of recipes it will run. Every half hour or so each node
+checks with the chef server and converges to the state which the chef cookbook
+specifies. Sysadmins can edit the attributes of a node on the chef server using
+the ``knife`` tool.
+
+Anatomy of a Coobook
+~~~~~~~~~~~~~~~~~~~~
+Most cookbooks begin with a table of contents, but Chef cookbooks are full of
+files. Here is a summary of important files in a cookbook.
+
+.. code:: text
+
+	.
+	├── attributes: A directory
+	├── Berksfile: Defines other cookbooks on which this one depends
```

Comment:
This defines **where to find** other cookbooks, not which ones the cookbook depends on.